server/src/test/java/org/elasticsearch/cluster/InternalClusterInfoServiceSchedulingTests.java

```diff
@@ -129,6 +132,44 @@ public void reroute(String reason, Priority priority, ActionListener<Void> liste
         );
         runUntilFlag(deterministicTaskQueue, becameMaster1);
 
+        // A node joins the cluster
+        {
+            Mockito.clearInvocations(mockEstimatedHeapUsageCollector, nodeUsageStatsForThreadPoolsCollector);
+            final int initialRequestCount = client.requestCount;
+            final AtomicBoolean nodeJoined = new AtomicBoolean();
+            clusterApplierService.onNewClusterState(
+                "node joins",
+                () -> ClusterState.builder(new ClusterName("cluster")).nodes(withJoiner).build(),
+                setFlagOnSuccess(nodeJoined)
+            );
+            // Don't use runUntilFlag because we don't want the scheduled task to run
+            deterministicTaskQueue.runAllRunnableTasks();
+            assertTrue(nodeJoined.get());
```

Comment:
this may be why the existing tests don't assert anything, it feels a little hacky, but this seemed like the right test to add to? Could also implement as an IT